activesupport/lib/active_support/messages/rotation_coordinator.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Messages
+    class RotationCoordinator # :nodoc:
+      def initialize(&secret_generator)
+        raise ArgumentError, "A secret generator block is required" unless secret_generator
+        @secret_generator = secret_generator
+        @rotate_options = []
+        @codecs = {}
+      end
+
+      def [](salt)
+        @codecs[salt] ||= build_with_rotations(salt)
+      end
+
+      def []=(salt, codec)
+        @codecs[salt] = codec
+      end
+
+      def rotate(**options)
+        changing_configuration!
+        @rotate_options << { secret_generator: @secret_generator, **options }
+        self
+      end
+
+      def rotate_defaults
+        rotate()
+      end
+
+      def clear_rotations
+        changing_configuration!
+        @rotate_options.clear
+        self
+      end
+
+      def on_rotation(&callback)
+        changing_configuration!
+        @on_rotation = callback
+      end
+
+      private
+        def changing_configuration!
+          if @codecs.any?
+            raise <<~MESSAGE
+              Cannot change #{self.class} configuration after it has already been applied.
+
+              The configuration has been applied with the following salts:
+              #{@codecs.keys.map { |salt| "- #{salt.inspect}" }.join("\n")}
+            MESSAGE
+          end
+        end
+
+        def build_with_rotations(salt)
+          raise "No options have been configured" if @rotate_options.empty?
```

Comment:
Any reason why we are requiring a rotation to be configured?

Reply:
The `MessageVerifiers` / `MessageEncryptors` are initialized with no options, and each set of options are added via [`RotationCoordinator#rotate`](https://github.com/rails/rails/pull/44179/files#diff-f7f8ca7150b620cf628f067821d9edfcf41bfbadb600cb76c67f44d3ff148928R21-R25).  So the first entry in `@rotate_options` will be the primary options required for signing / encrypting, and the rest of the entries will be fallback options for verification / decryption.

We could configure and store the primary options in a different way (e.g. via `initialize`), but that would make the API more complicated, particularly when [`clear_rotations`](https://github.com/rails/rails/pull/44179/files#diff-f7f8ca7150b620cf628f067821d9edfcf41bfbadb600cb76c67f44d3ff148928R31-R35) is involved.

(I felt it was important to have a "reset" method like `clear_rotations`, so that Rails can provide a pre-configured `Rails.application.message_verifiers`, but users can completely reconfigure it.)

I debated the names for `@rotate_options` and `clear_rotations`.  In the end, I felt like those names were most consistent with the `RotationCoordinator#rotate` API.  But I am open to different names, such as `@configuration` and `clear_configuration`.  We could also rename `RotationCoordinator#rotate`, if that helps.

Reply:
I think this answers my request. Thanks!